tools/boardgen.py: Add a common library for ports to implement make-pins.py.

jimmo opened the pull request:
Each port currently has it's own complete implementation of make-pins.py. Historically the original implementation was done for stm32, and this has been copied and modified to 6 other ports (cc3200, teensy, nrf, mimxrt, rp2, renesas-ra), and then a brand-new from-scratch implementation was recently added for esp32.

They all do roughly the same things, but with some differences (e.g. hidden pins, output format, naming, argument parsing), and in some places there is unused code that hasn't been removed while adapting. Importantly, they duplicate a lot of code, and for someone implementing a new port it's difficult to see exactly the bits that are port-specific and the bits that don't need to be modified, and it's easy to make mistakes. I am in the process of adding a make-pins.py for the Zephyr port, which is why I started looking into this.

This PR does the following things:
 - Small fixes to the port makefiles to simplify the make-pins.py rules.
 - Remove the unnecessary generation of pin_qstr.h (pins.c is found in qstr discovery anyway), so this is not needed.
 - Remove the generation of pins_af.py (and the corresponding example that uses it). Not all ports supported this and it's not worth the complexity.
 - Make all versions of make-pins.py take the same common command line arguments, and output the same way (always to files, not to stdout).
 - Simplification to esp32 to make the irq object a subfield of pin.
 - Minor optimisation to rp2 to make the pin table contain the actual pins rather than pointers to pins.
 - Fix some errors in stm32 pin.csv and af.csv files.

And then the main work in this PR is to implement `tools/boardgen.py` which is a helper for ports to implement `make-pins.py`. It's named `boardgen` because I hope it can be expanded to do more than just pins. It supports numeric and named pins, as well as optional alternate function support.

Then the esp32 make-pins is updated to use this (chosen because it's the simplest make-pins.py), then rp2 (because it's quite similar to esp32, just with alternate functions), and then stm32 (because it's the most complex), and then the remaining ports: samd, mimxrt, renesas-ra. I have left out nrf, teensy, cc3200.

For stm32, the new version is a bit stricter so this uncovered a few issues with board pins.csv and af.csv files.

_This work was funded through GitHub Sponsors._

jimmo committed:
rp2/boards/make-pins.py: Don't generate qstrs.

Also remove af-const header, as this is left over from the STM32 version
and unused.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
examples/pins.py: Remove this pins printing example.

It's not supported on all ports, adds complexity to the build to generate
pins_af.py, and can mostly be replicated just by printing the pin objects.

Remove support for generating pins_af.py from all ports (nrf, stm32,
renesas-ra, mimxrt, rp2).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32/boards/make-pins.py: Don't generate qstrs.

Qstrs are picked up from the generated pin source files in the usual qstr
processing stage.

Similarly for the stm constant qstrs.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
nrf/boards/make-pins.py: Don't generate qstrs.

The output pins.c can be processed for qstrs like any other C file.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
renesas-ra/boards/make-pins.py: Don't generate qstrs.

The output pins.c can be processed for qstrs like any other C file.

Also remove af_const from Makefile (unimplemented in make-pins.py) and fix
target dependency on ad_const.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/mkrules.mk: Add rule for compiling auto-generated source files.

This prevents each port Makefile from having to add an explicit rule for
`build-BOARD/pins_BOARD.c`.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
ports: Remove SRC_QSTR_AUTO_DEPS from all ports' Makefiles.

It's unused.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
ports: Standardise arguments and output for make-pins.py script.

All ports now use `--board-csv`, `--prefix`, `--output-souce`,
`--output-header` and no longer write to stdout.  This matches the esp32
implementation.

Ports that have an AF input use `--af-csv` (to match `--board-csv`).

Any additional output files are now prefixed with `output-` (e.g.
`--output-af-const`).

Default arguments are removed (all makefiles should always specify all
arguments, using default values is likely an error).

Replaced the `af-defs-cmp-strings` and `hdr-obj-decls` args for stm32 with
just `mboot-mode`.  Previously they were set on the regular build, now the
logic is reversed so mboot sets it.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
cc3200/boards/make-pins.py: Don't generate qstrs.

The output pins.c can be processed for qstrs like any other C file.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp32/machine_pin: Make irq object a sub-field of pin object.

This simplifies pin generation.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/boardgen.py: Add initial implementation of a common make-pins.py.

For now, this implements the functionality required for esp32 and rp2,
including support for numeric pins, rp2 alternate functions, and rp2
extended pins.

This also updates the rp2 port to use the same structure for pins.h and
pins.csv as for esp32, and moves the pin definitions directly into the
table (rather than having a table of pointers), which is a small code size
improvement.

Support for "hidden" pins in pins.csv is added (matching the stm32
implementation).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32/boards/make-pins.py: Update to use tools/boardgen.py.

Requires additions to tools/boardgen.py for stm32 pin generation.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32: Rename pin_obj_t to machine_pin_obj_t.

This is now consistent with other ports.

Also renamed `pin_{board/cpu}_pins_locals_dict` to
`machine_pin_{board/cpu}_pins_locals_dict`.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32/boards: Fix errors in pins.csv and af.csv.

Fixes are:
- Comment out lines in pins.csv that do not have valid CPU pins.
  It's useful to keep these in the file as "documentation" but in order to
  make make-pins.py stricter they need to be commented out.
- Fix some typos (missing P prefix) in pins.csv.
  This resulted in some missing board pins.
- Fix some typos in af.csv files.
  Some typos of "ADC" and some other that were previously ignored.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

robert-hh commented:
Yes, this make_pins stuff is pretty feral. 
I have a unpublished PR for SAMD which swaps the fields in pins.csv, which is different ATM to the other ports. I can push that now, such that it is at least visible.

robert-hh commented:
@jimmo. I added a commit to PR #12212 to match the argument names of the make-pins.py equivalent to your PR here, and also renamed that script to make-pins.py.

jimmo commented:
> @jimmo. I added a commit to PR #12212 to match the argument names of the make-pins.py equivalent to your PR here, and also renamed that script to make-pins.py.

Thanks for that @robert-hh , I will integrate that into this PR.

One question with SAMD, I notice the convention in this port for cpu pins up to 10 appears to be `PA01` whereas on STM32 that would be `PA1` (i.e. samd is adding the leading zero).

In general I think we should match the style of what the vendor calls them, but also I would like where possible for things to be as consistent as possible across ports, and of course I'm hesitant to break compatibility here. What should we do?

robert-hh commented:
> I will integrate that into this PR.

This commit must be used with the first commit in the PR, which swaps the columns of the pins.csv files. Otherwise the builds will fail.

>  I notice the convention in this port for cpu pins up to 10 appears to be PA01

MicroChip uses consistently the 2-digit numbering for pins in their documents and library. See e.g. chapter 32.5.1 of the SAMD51 data sheet: 
"_The I/O lines of the PORT are mapped to pins of the physical device. The following naming scheme is used:
Each line bundle with up to 32 lines is assigned an identifier 'xy', with letter x=A, B, C… and two-digit number y=00, 01, …31. Examples: A24, C03.
PORT pins are labeled 'Pxy' accordingly, for example PA24, PC03. This identifies each pin in the device uniquely._"
The board vendors do the same in their schematics, when they denote MCU pins. So I think we should stick to it.

jimmo commented:
> So I think we should stick to it.

OK, makes sense to me! Thanks @robert-hh

jimmo committed:
samd/boards/make-pins.py: Update to use tools/boardgen.py.

This replaces the previous make-pin-table.py with an implementation based
on boardgen.py.

- MICROPY_PY_MACHINE_PIN_BOARD_CPU macro is removed. This isn't optional
  on other ports, so no need for it to be optional on SAMD.
- pin_af_table is removed, and lookups just search the cpu dict instead
  (this saves N*wordsize bytes of firmware size to have this extra table).
- pins.csv is now BOARD,CPU to match other ports.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
mimxrt/boards/make-pins.py: Update to use tools/boardgen.py.

Minor change to remove support for using numeric IDs for machine.Pin.  This
was previously based on the index of the pin in the board csv, but this is
different (and incompatible) with other ports.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
@robert-hh I've updated this PR with samd and mimxrt using boardgen.py now.

robert-hh commented:
I was about to close this PR, since two of the commits are already covered by #12211. And, as you said, being able to omit Pin.cpu is kind of inconsistent, even it Pin.cpu is not available at all ports.

robert-hh commented:
Getting lost between 2 PRs. What I wanted to say:

You digged up quite a bit of the code, which was grown over time and may have needed some clean-up. From a distance, things look sometimes more clear. I made a trial build for a SAMD51 & SAMD21 board and noticed, that the the element `name` in machine_pin_obj_t is not properly assigned any more. It used to be the board pin name or '-', if no name is assigned. Now it is the CPU pin name, meaning that the function pin_name() returns the wrong value.

robert-hh commented:
Teensy 4.1 and MIMXRT1176_EVK do not build. And OLIMEX_RT1010 and ADAFRUIT_METRO_M7 still have the UART RTS/CTS bug (not your playground).

jimmo commented:
> the element name in machine_pin_obj_t is not properly assigned any more. It used to be the board pin name or '-', if no name is assigned. Now it is the CPU pin name, meaning that the function pin_name() returns the wrong value.

That was intentional, and it's how it behaves on other ports. e.g. on stm32:

```py
>>> machine.Pin.board.LED_RED
Pin(Pin.cpu.B1, mode=Pin.OUT)
```

I've updated the samd printing to match the same style. FYI you can use `%q` as a format specifier with mp_printf to directly print a qstr, so the pin_name function converting to `char*` wasn't necessary.

I agree it's neat that the pin knew its board name, but I'd rather make this consistent.

> Teensy 4.1, SEEED_ARCH_MIX and MIMXRT1176_EVK do not build.

The way pins.csv works on other ports is that unless a pin is in pins.csv it doesn't get generated. (With the exception of rp2 and esp32 which always generate the fixed set of numeric cpu pins).

But to keep "pin to board" mapping centralised, this is also supported by pins.csv even for pins that will not be made available to Python. This is why the stm32 pins.csv (and now everything using boardgen.py) supports "hidden" pins. The idea is that you can add, e.g.

```
-ENET_RESET,-GPIO_LPSR_12
```

to pins.csv to make `pin_ENET_RESET` available (pointing to pin_GPIO_LPSR_12_obj), but neither will be in Pin.board/Pin.cpu. So this means `eth.c` can use `pin_ENET_RESET`, removing the defines in mpconfigboard.h.

Anyway, I had already fixed this for the other boards, but missed TEENSY41 and missed the 1G eth pins for MIMXRT1176_EVK.

Similarly for the USDHC1 pins, these also needed to be added to pins.csv.

I've patched in the changes from the RTS/CTS PR and now all boards are compiling.

robert-hh commented:
> That was intentional, and it's how it behaves on other ports. e.g. on stm32:

I do not agree to that change. The name field is intended to show the board pin name, so that the Pin print shows both the board name and the CPU name. Otherwise the field itself is obsolete, because the cpu name and the pin id are directly related. And since it's more likely that people use the board name for pins, it should be shown when devices properties are printed.

robert-hh commented:
The changes for the MIMXRT boards seems fine. The three boards that did not build before build and Ethernet works. The info about the hidden pins is useful.

> I agree it's neat that the pin knew its board name

From a usage perspective it's essential, not neat. People should be able to tell the pin name that is shown by the a board API without looking into information which exists somewhere.

robert-hh commented:
Another thing that is missing: when testing with the SAMD port i noticed, that pins which are not assigned to a board name are missing in the Pin.cpu list. At the moment they are present and it is useful to have them.

jimmo commented:
> Pin.cpu must contain all Pins of the MCU. The respective pin_af.csv tables list all possible pins. So that must be used for creating Pin.cpu.

This change is also for compatibility with the other ports. I realise that there isn't an easily definable "common behavior" because there is so much variation in each of the make-pins.py implementations, so to be clear, wherever there's a discrepancy I'm using stm32 as the reference implementation.

The idea is that if a pin isn't in the board pins.csv, then it's not "on the board" and therefore there would be no way to access it and no reason to make it available. "on the board" doesn't just mean "available on a header" though, it's more that the pin is routed _somewhere_ on the board and has some function.

And when it comes to making it available, the "hidden" pins feature allows it to be just made available to C (e.g. the ETH_RESET pins discussed earlier are a good example). CPU pins also don't have to be given board aliases -- e.g. `,PA1` is a valid row in pins.csv.

One obvious reason for doing this is code size -- there's no point generating `machine_pin_obj_t` instances in ROM for every possible CPU pin if we know it's not used anywhere.

So with the example you ran into with SAMD testing, the intention is that the pin should have been in pins.csv (possibly without a board name). Is that reasonable? Or is there another scenario I'm missing?

FWIW, on stm32, not all board definitions get this right, so some clean-up here is in order too. For example there are many boards with pins.csv that just list every possible pin as `PAxy,PAxy`, which is pointless (adding the CPU names to Pin.board). Although many boards do list the CPU name on the silkscreen...so you could argue this is correct in some cases (but I think when a pin is clearly the CPU name, it should not be in Pin.board).

There are two exceptions to this on existing ports -- esp32 and rp2, where we added named pin support after board definition support, and so we have boards that haven't been updated with pins.csv files. So for now all CPU pins are enabled. But this should be fixed once we have pins.csv for all boards. ESP32 also doesn't have the Pin.cpu dict at all.

I really feel that consistency is the most important thing. Once everything is using boardgen.py, we can make changes as necessary in one place that apply to all ports.

robert-hh commented:
>  "on the board" doesn't just mean "available on a header" though, it's more that the pin is routed somewhere on the board and has some function.

I verified for the SAMD and MIMXRT port that one can simply add all pins from the pin_af.csv file to pins.csv and get back a complete list of usable CPU pins into Pin.cpu. Usable CPU pins could both be pins that are routed to the board, or pins that are just internal to the MCU package but have a function, like those connected to an in-package peripheral.  Adding that to each pins.csv file is possible, but lacks elegance. An option would be an include mechanism or a addition cpu_pins.csv file for common subsets. 

> there's no point generating machine_pin_obj_t instances in ROM for every possible CPU pin if we know it's not used anywhere.

How can you know, what users will use, how users will use boards?  We should not patronize users.

> wherever there's a discrepancy I'm using stm32 as the reference implementation.

The STM32 port has many excellent features implemented. But it carries with it a lot of legacy and history, that may not all be useful for getting a new common structure. I prefer to use the RP2 port as a reference. It is way more consistent and clear. Obviously it's for a single MCU only, which makes it easier to read and maintain.

> Once everything is using boardgen.py, we can make changes as necessary in one place that apply to all ports.

I do not like to have a state which breaks things that worked before, just to eventually bring them back sometime later, if a PR for it is made and if it will be processed.

andrewleech commented:
> > there's no point generating machine_pin_obj_t instances in ROM for every possible CPU pin if we know it's not used anywhere.
> 
> How can you know, what users will use, how users will use boards? We should not patronize users.
> 

In my understanding of this there's a difference between what a CPU can provide and what a user can physically get access to. 

A solid example that comes to mind is the [Seeed Wio-E5 Wireless Module](https://www.seeedstudio.com/LoRa-E5-Wireless-Module-p-4745.html), from a micropython point of view this is just a STM32WL chip and SX126X Lora radio. It's a sealed can module, with only a small subset of the STM pins routed to the outside world and internal lora transceiver. 
I definitely think it makes sense on a module (or board using this module) to only expose pins that are routed out of the module / used internally for radio, any others are just consfusing noise taking up flash space.

On a more generic front, if a board only has some CPU pins routed from the chip to peripherals / headers / test-points with the rest not even having tracks laid, I'd prefer to not have those blank pins included by default. 
If I'm going to take a board like that and dead-bug wire some other parts onto those cpu pins, well I'm also happy to custom compile my firmware as it's no longer matching that board definition.

robert-hh commented:
> In my understanding of this there's a difference between what a CPU can provide and what a user can physically get access to.

Physical access is not the only aspect. It should include the aspect of what user code can use, even if a CPU pin is not routed on or to the board. Two examples:
- At the SEEED XIAO SAMD21 the PA03 is not routed to the board edge. PA03 is the AREF pin, and kept floating. That makes using the ADC a little bit tricky. But since it is available as CPU pin, one set set it to high level and have a AREF voltage applied at the pin.
- The MIMXRT PWM allows to synchronize PWM channels to another specific channel of a PWM module. These specific channels may not be routed to a board connector, but that's not required to use the sync mechanism. They must just be available to the code as Pin object.
The may be more examples of situation, where a CPU pin can be used even if not routed to the board connectors or used on the board. Not all people a willing or able to build their own firmware in case they miss a pin definition. So why make things more complicated and restrictive? Pin.board may list only the labelled pins and dedicated pins on the board. But Pin.cpu should have all that can be used by code.

andrewleech commented:
Ah, that is a compelling argument... yes there are times you may use a physical pin for "connecting" two internal peripherals together, such as MCO clock output on stm being re-used as a clock input for a different module.

jimmo commented:
Updated the cc3200 make-pins.py.

It appears that this didn't previously worked and only emitted cpu pins (but called them board pins), and the board names in pins.csv was ignored. As far as I can tell it's doing the right thing now. You can construct a pin using `Pin("CPUNAME")` or `Pin("BOARDNAME")` as well as using `Pin.board.BOARDNAME`.

jimmo commented:
@robert-hh @andrewleech I have updated mimxrt and samd to make all pins from the af csv file available as CPU pins. (It's a one line change and can be done on a per-port basis in make-pins.py -- the pin can set `self._available` in `add_af`). I have verified that this now generates the same set of pins as before. The board.csv file may still choose to hide the cpu or board name from Python.

I really wanted to use this effort to improve consistency across the ports, but honestly I'm very happy as long as any inconsistencies are clear and deliberate. I hope that the new simpler make-pins.py files have a much higher ratio of "port-specific" content to them, so it's clear what they do and why, and not lost in the noise of the common functionality (i.e. generating output and parsing of files).

robert-hh commented:
Thank you very much. It works.
I had already prepared a commit of adding these pins again as unnamed pins to pins.csv. But that's not needed any more.

robert-hh commented:
> Updated the cc3200 make-pins.py.

I have built a CC3200 image for WiPy. There is no Pin.cpu class, just a Pin.board class. The Pin.cpu dictionary is created, but the class & type definitions in pybpin.c and pybpin.h are missing.
With Pin.board, the labeling is surprising. The board labels the pins at the package as GP0 - GP28, while Pin.board show Lx and Rx labels.

jimmo committed:
renesas-ra/boards/make-pins.py: Update to use tools/boardgen.py.

This removes previously unused functionality to generate pins_ad_const.h,
as well as the unused handling of pin AF in machine_pin.c.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> I have built a CC3200 image for WiPy. There is no Pin.cpu class, just a Pin.board class. The Pin.cpu dictionary is created, but the class & type definitions in pybpin.c and pybpin.h are missing.

Thanks for testing! Yeah, it didn't previously have a Pin.cpu dict, but seemed to be using the board dict as if it were the cpu dict. I could add Pin.cpu to make it available to Python (but instead I just made it so Pin("X") supported the cpu names also.

> With Pin.board, the labeling is surprising. The board labels the pins at the package as GP0 - GP28, while Pin.board show Lx and Rx labels.

I was about to suggest that I should update pins.csv, which was clearly not only previously unused/ignored but also wrong.

But instead I'm going to revert this commit and not touch the CC3200 port.

> P.S.: I am never sure if the CC3200 port really should be updated. The board went out of sale since long. Maybe with version 2.x it can be dropped.

I don't think there's any reason why it needs to be tied to v2.x. Similarly I already had no intention of updating make-pins.py for Teensy.

robert-hh commented:
> only previously unused/ignored but also wrong.

I should have looked with `git blame` at that file. It has been made 2015 by Daniel Campora.

> But instead I'm going to revert this commit and not touch the CC3200 port.

That seems reasonable, as long as Pin("Gnn") works. The Teensy 3.2 port was a starting point, which did not evolve further. There were also trial branches for other members the Kinetis chip family (Teensy 3.5 and 3.6), but that one was closed long ago.

jimmo committed:
{cc3200,nrf}/boards/make-pins.py: Add a note about tools/boardgen.py.

It's not worth the effort to update these ports to use boardgen.py, but
put a note just in case anyone uses this as a reference for a new port.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
And the last one... renesas-ra is now done. Found a couple of minor issues in the af.csv parsing which now means a few pins that were missed before are now available. Also removed a bunch of unused code.

dpgeorge commented on ports/esp32/boards/make-pins.py:
```diff
@@ -1,185 +1,67 @@
 
-MAX_CPU_PINS = 49
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "../../../tools"))
+import boardgen
```
This import of boardgen appears in the wrong commit.

dpgeorge commented on ports/stm32/boards/make-pins.py:
```diff
@@ -1,59 +1,17 @@
-Generates pin source files based on an MCU alternate-function definition (eg
-stm32f405_af.csv) and a board-specific pin definition file, pins.csv.
 
-The pins.csv file can contain empty lines, comments (a line beginning with "#")
```
Would be good to keep this documentation about the format of pins.csv.  Maybe at the top of `tools/boardgen.py`?

dpgeorge commented:
The mimxrt and rp2 ports have definitions of `GEN_PINS_AF_CONST` in their makefile/cmakelists that are no longer needed.

dpgeorge commented on ports/stm32/mboot/Makefile:
```diff
@@ -245,10 +243,10 @@ $(GEN_ROOT_POINTERS): | $(HEADER_BUILD)
+	$(Q)$(PYTHON) $(MAKE_PINS) --board-csv $(BOARD_PINS) --af-csv ../$(AF_FILE) --prefix $(PREFIX_FILE) \
+		--output-source $(GEN_PINS_SRC) --output-header $(GEN_PINS_HDR) \
+	  --output-af-const $(GEN_PINS_AF_CONST) --output-af-defs $(GEN_PINS_AF_DEFS) \
+	  --mboot-mode
```
probably should be indented to match above lines

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -698,10 +688,10 @@ $(GEN_PLLFREQTABLE_HDR): $(PLLVALUES) | $(HEADER_BUILD)
-# make both modstm_const.h and modstm_qstr.h
-$(HEADER_BUILD)/%_const.h $(BUILD)/%_qstr.h: $(CMSIS_MCU_HDR) make-stmconst.py | $(HEADER_BUILD)
+# make both modstm_const.h
+$(HEADER_BUILD)/%_const.h: $(CMSIS_MCU_HDR) make-stmconst.py | $(HEADER_BUILD)
```
This comment about "both" is no longer valid, and maybe this rule can just become:
```
$(HEADER_BUILD)/modstm_const.h: $(CMSIS_MCU_HDR) make-stmconst.py | $(HEADER_BUILD)
```

dpgeorge commented on tools/boardgen.py:
```diff
@@ -0,0 +1,517 @@
+        # Optionally load pins.csv to get cpu->board name mappings.
+        if self._enable_af and self.args.af_csv:
+            print("// --af-csv {:s}".format(self.args.af_csv), file=out_source)
+            self.parse_af_csv(self.args.af_csv)
```
The original esp32 version of make-pins.py had the parsing of input files separated from the output printing.  First it parsed, then it printed.  But here those functions are mixed.

Don't you think it's cleaner to parse everything first, then print?  They can be separate functions/methods in this PinGenerator.

jimmo replied:
> Don't you think it's cleaner to parse everything first, then print?

Honestly, not really? It's printing the information about the loading, not printing while parsing.

The alternative is more complicated -- keeping track of the filenames etc to print them later.

Even the original esp32 version did the prefix the same way it is here.

dpgeorge replied:
OK, keep it as you have it.

dpgeorge commented:
There's a comment in the commit message for commit `tools/boardgen.py: Initial implementation of a common make-pins.py.` which says:

    The AF handling for rp2 was also simplified to just use a static table in
    make-pins.py, removing the complexity of loading a csv file that is
    identical for all boards.

I think this comment is wrong and should just be removed from that commit message.

dpgeorge commented on ports/stm32/boards/STM32F439/pins.csv:
```diff
@@ -1,4 +1,4 @@
-PC0,PC0
+FPC0,PC0
```
Why does this need to be changed?

dpgeorge commented on ports/stm32/boards/stm32h7b3_af.csv:
```diff
@@ -1,15 +1,13 @@
 Port,,AF0,AF1,AF2,AF3,AF4,AF5,AF6,AF7,AF8,AF9,AF10,AF11,AF12,AF13,AF14,AF15,
 ,,AF0,AF1,AF2,AF3,AF4,AF5,AF6,AF7,AF8,AF9,AF10,AF11,AF12,AF13,AF14,AF15,ADC
 PortA,PA0,,TIM2_CH1/TIM2_ETR,TIM5_CH1,TIM8_ETR,TIM15_BKIN,I2S6_WS/SPI6_NSS,,USART2_CTS/USART2_NSS,UART4_TX,SDMMC2_CMD,SAI2_SD_B,,,,,EVENTOUT/EVENTOUT,ADC1_INP16,
-PortA,PA0_C,,TIM2_CH1/TIM2_ETR,TIM5_CH1,TIM8_ETR,TIM15_BKIN,I2S6_WS/SPI6_NSS,,USART2_CTS/USART2_NSS,UART4_TX,SDMMC2_CMD,SAI2_SD_B,,,,,EVENTOUT/EVENTOUT,ADC1_INP16,
```
These pins ending in _C are actually physical pins, used for dedicated ADC inputs.  But since we don't support them I guess it's OK to remove them from this file.

dpgeorge commented on ports/stm32/boards/stm32h7b3_af.csv:
```diff
@@ -68,7 +64,6 @@ PortD,PD6,,,SAI1_D1,DFSDM1_CKIN4,DFSDM1_DATIN1,I2S3_SDO/SPI3_MOSI,SAI1_SD_A,USAR
 PortD,PD7,,,,DFSDM1_DATIN4,,I2S1_SDO/SPI1_MOSI,DFSDM1_CKIN1,USART2_CK,,SPDIFRX_IN1,OCTOSPIM_P1_IO7,SDMMC2_CMD,FMC_NE1,,,EVENTOUT/EVENTOUT,,
 PortD,PD8,,,,DFSDM1_CKIN3,,,,USART3_TX,,SPDIFRX_IN2,,,FMC_D13/FMC_DA13,,,EVENTOUT/EVENTOUT,,
 PortD,PD9,,,,DFSDM1_DATIN3,,,,USART3_RX,,,,,FMC_D14/FMC_DA14,,,EVENTOUT/EVENTOUT,,
-PortD,PDR_ON,,,,,,,,,,,,,,,,EVENTOUT,,
```
This is also a physical pin, although I don't know how to use it... so removing it from here seems OK.

dpgeorge commented on ports/stm32/cyw43_configport.h:
```diff
@@ -80,7 +80,7 @@
 #define cyw43_hal_ticks_ms              mp_hal_ticks_ms
 
-#define cyw43_hal_pin_obj_t             mp_hal_pin_obj_t
+#define cyw43_hal_machine_pin_obj_t             mp_hal_pin_obj_t
```
Please retain indenting style.

dpgeorge replied:
Actually, this line and the other one in this file should not change at all.

jimmo commented on tools/boardgen.py:
```diff
@@ -0,0 +1,517 @@
+# This file is part of the MicroPython project, http://micropython.org/
```
It's a library contain the "PinGenerator", whereas a script called "gen_board.py" would imply that it's a script that generates a board (which is actually the role of `make-pins.py`).

I do not love `boardgen.py` though.

andrewleech commented on tools/boardgen.py:
```diff
@@ -0,0 +1,556 @@
+                            # If this is the first header row then initialise
+                            # the headings dict.
+                            for af_idx, header in enumerate(row[af_col:]):
+                                headings[af_idx] = header.strip()
```
re: stm and missing AF columns, some chips don't have/use certain AF domains, eg F413 jumps from AF12 to AF15. Similarly, with a "fake" AF16 added for ADC, chips like F405 then jump from AF13 to AF16.
This code based on enumerate will give incorrect numbers to columns after any skipped columns if they're not present in the CSV. 

It would be fine to enforce csv files to always include all domains, though to avoid weird errors it would also be beneficial here to check the column header actually includes the expected number, or parse the number from the header rather than enumerate?
enforce:
```
if str(af_idx) not in header:
    raise PinGeneratorError(
        "Out of sequence AF heading {}".format(header)
    )
```

parse:
```
                            for column, header in enumerate(row[af_col:]):
                                af_idx = int(header.strip().replace("AF", ""))
                                headings[column] = (af_idx, header.strip())
```
...
```
                    for col, af in enumerate(row[af_col:]):
                        af = af.strip()
                        if not af:
                            continue
                        af_idx, heading = headings.get(col, (col, ""))
                        pin.add_af(af_idx, heading, af)
```

jimmo replied:
I've added some commits to format all the af files consistently (as well as padding them as discussed elsewhere), and make-pins.py now checks the headers.

iabdalkader commented:
Hi!
Please don't remove the stm32 dual pad pins from CSV files, I have a PR here to add support for them:
https://github.com/micropython/micropython/pull/12706

jimmo committed:
stm32/boards: Format stm32 alternate function csv files.

Changes are:
- Pad all cells to make them easier to read.
- Ensure all files have exactly 19 columns (Port,Pin,AF0-15,ADC)

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32/boards/make-pins.py: Add initial support for H7 dual-pad pins.

This fixes the H7 af.csv files to include the dual-pad information, by
listing the ADCs supported on the _C pad with a C_ADC prefix.

Minimal change to make-pins.py to ignore these entries. This will be
implemented later to emit constants (similar to ADC.CORE_TEMP) to access
these channels.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> Hi! Please don't remove the stm32 dual pad pins from CSV files, I have a PR here to add support for them: #12706

I've added the C_ADC entries to the af.csv files as discussed in #12706.

iabdalkader commented:
> I've added the C_ADC entries to the af.csv files as discussed in #12706.

@jimmo 

> Damien suggested that rather than waiting for ADCBlock to exist in order to access the _C channels, today we could add e.g. ADC(ADC.PC2_C) to achieve this.

I left a comment on #12706. This feels like a workaround to access those channels, my goal is not just access the extra ADC channels, the main goal is to support mapping "board pins" to `Pxy_C` pads. For example on Portenta board pin `A0` is connected to `PA0_C` so the following should work:

```Python
adc = ADC("A0") # ADC12_IN0 channel
```

It's important to use `A0` because it makes examples portable (on Giga board pin A0 is PC4) `ADC(ADC.PC2_C)` is Not portable. After the changes here, can we still support that ?

NOTE: This fact that `A0` is connected `PC2_C` is only documented in schematics and datasheets, the user only sees `A0` on the board's silkscreen. How would a user know which pad maps to an `Ax` pin ?

jimmo commented:
Let's keep the discussion about the ADCs in #12706. I've replicated @iabdalkader's comment above there.

dpgeorge commented:
Merged!

Any remaining inconsistencies/regressions can be dealt with separately.
